content/advent-2018/go-arrow.md

```diff
@@ -0,0 +1,531 @@
+
+So why would anyone use Go for this?
+
+Well... the usual reasons:
+
+- Go is easier to deploy,
+- Go is faster than pure Python,
+- Go code is simpler when dealing with concurrency,
+- Go code and results are easy to reproduce,
+- Go code is more amenable to mechanical refactoring,
+- Go code tends to be more robust at scale than Python code.
+
+The real question is: do we have all the building blocks, all the libraries to write a modern, robust, efficient analysis pipeline in Go?
+
+Most of the science-y things one might carry out with the Python scientific stack (NumPy/SciPy) can be performed with [Gonum](https://gonum.org):
+
+- linear algebra, matrices,
+- differentiation,
+- statistics, probability distributions,
+- integration, optimization,
+- network creation,
+- Fourier transformations,
+- plots.
+
+There are also a few packages that enable file-level interoperability with many useful data science oriented or scientific "file formats":
+
+- [CSV](https://godoc.org/encoding/csv)
+- [SQL](https://godoc.org/database/sql)
+- [Numpy data files](https://godoc.org/github.com/sbinet/npyio)
+- [HDF5](https://godoc.org/gonum.org/v1/hdf5)
+- [FITS](https://godoc.org/github.com/astrogo/fitsio)
+- [HDFS](https://godoc.org/github.com/colinmarc/hdfs)
+- [Parquet](https://godoc.org/github.com/xitongsys/parquet-go/parquet)
+- [ORC](https://godoc.org/github.com/scritchley/orc)
+- [Cassandra](https://godoc.org/github.com/gocql/gocql)
+- ...
+
+Many basic ingredients are already there, even if some are still missing.
+
+But even if basic file-level interoperability is (somewhat) achieved, one still needs to implement readers, writers and converters from one file format to any another when integrating with a new analysis pipeline:
+
+![file formats](/postimages/advent-2018/go-arrow/arrow-copy.png)
+
+In many scenarios, this implies a large fraction of the computation is wasted on serialization and deserialization, reimplementing over and over the same features for converting from `format-1` to `format-n` or from `format-1` to `format-m`, _etc..._
+
+What if we had a common data layer between all these analysis pipelines, a common tool that would also be efficient?
+
+## Arrow
+
+What is [Arrow](https://arrow.apache.org)? From the website:
+
+```
+Apache Arrow is a cross-language development platform for in-memory data.
+
+It specifies a standardized language-independent columnar memory format for flat
+and hierarchical data, organized for efficient analytic operations on modern
+hardware.
+It also provides computational libraries and zero-copy streaming messaging and
+interprocess communication.
+```
+
+![arrow shared](/postimages/advent-2018/go-arrow/arrow-shared.png)
+
+The idea behind Arrow is to describe and implement a cross-language mechanism for efficiently sharing data across languages and processes.
+This is embodied as an `arrow::array` in C++, the reference implementation.
```

Comment:
`arrow::Array`. In fairness, the Java library is also a reference implementation, even if its development activity is a bit lower than C++. Might be worth putting them on equal footing (even though the Java library uses the term "vector" instead of "array", confusingly)